crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,320 @@
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    pub(crate) fn get_sealed_headers<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<SealedBlockHeader>>>
+    where
+        V: P2pDb,
+    {
+        let mut headers = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(header) = self.sealed_block_headers.get(&block_height) {
+                headers.push(header.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
+
+        if missing_ranges.is_empty() {
+            self.update_metrics(increment_p2p_req_res_cache_hits);
+            return Ok(Some(headers))
+        }
+
+        self.update_metrics(increment_p2p_req_res_cache_misses);
+        let missing_headers = view.get_sealed_headers(missing_ranges.clone())?;
+        if let Some(missing_headers) = &missing_headers {
+            for header in missing_headers.iter() {
+                self.sealed_block_headers
+                    .insert((*header.entity.height()).into(), header.clone());
+                headers.push(header.clone());
+            }
+        }
+        Ok(missing_headers)
+    }
+
+    pub(crate) fn get_transactions<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<Transactions>>>
+    where
+        V: P2pDb,
+    {
+        let mut transactions = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(cached_tx) = self.transactions_on_blocks.get(&block_height) {
+                transactions.push(cached_tx.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
+
+        if missing_ranges.is_empty() {
+            self.update_metrics(increment_p2p_req_res_cache_hits);
+            return Ok(Some(transactions))
+        }
+
+        self.update_metrics(increment_p2p_req_res_cache_misses);
+        let transactions = view.get_transactions(missing_ranges.clone())?;
+        if let Some(transactions) = &transactions {
+            for (block_height, transactions_per_block) in
+                missing_ranges.zip(transactions.iter())
+            {
+                self.transactions_on_blocks
+                    .insert(block_height, transactions_per_block.clone());
+            }
+        }
+        Ok(transactions)
```

Comment:
Nit: This code is very similar to the one in `get_sealed_headers`. Perhaps it would be nice to factor it out to a separate helper function. Though it's only repeated once so if the helper function gets overly complex with generics and function pointers (e.g. for the metrics functions) then feel free to keep it as-is.

Reply:
i agree about refactoring it, but its more readable this way. if it's required in the future, (i.e serving more types of data over req/res), we can make it generic.

Reply:
Makes sense :+1: